src/ctx_macro.rs

```diff
@@ -0,0 +1,239 @@
+        for ($ctx, ($len, $offset)) in core::iter::zip([$($ctx_expr,)*], core::iter::zip([$($len_expr,)*], [$($offset_expr,)*])) {
+            case_set!($UP_TO, $($WITH_DEFAULT,)? $ctx=$ctx, $len=$len, $offset=$offset, $body);
+        }
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident, $offset:ident, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len, $offset=$offset, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident=$len_expr:expr, $offset:ident, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len_expr, $offset=$offset, { $(($buf, $val)),* });
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident, $offset:ident=$offset_expr:expr, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len, $offset=$offset_expr, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len_expr, $offset=$offset_expr, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $ctx:ident=$ctx_expr:expr, $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
+        let $ctx = $ctx_expr;
+        let $len = $len_expr;
+        let $offset = $offset_expr;
+        {
+            #[allow(unused_macros)]
+            macro_rules! set {
+                ($buf:expr, $val:expr) => {
+                    assert!($offset <= $buf.len() && $offset + $len <= $buf.len());
+                };
+            }
+            #[allow(unused_macros)]
+            macro_rules! set_disjoint {
+                ($buf:expr, $val:expr) => {
+                    assert!($offset <= $buf.len() && $offset + $len <= $buf.len());
+                };
+            }
+            $body
+        }
+        macro_rules! exec_block {
+            ($N:literal, $block:block) => {
+                {
+                    #[allow(unused_macros)]
+                    macro_rules! set {
+                        ($buf:expr, $val:expr) => {
+                            // SAFETY: The offset and length are checked by the
+                            // assert outside of the match.
+                            let buf_range = unsafe {
+                                $buf.get_unchecked_mut($offset..$offset+$N)
+                            };
+                            *<&mut [_; $N]>::try_from(buf_range).unwrap() = [$val; $N];
+                        };
+                    }
+                    #[allow(unused_macros)]
+                    macro_rules! set_disjoint {
+                        ($buf:expr, $val:expr) => {
+                            // SAFETY: The offset and length are checked by the
+                            // assert outside of the match.
+                            let mut buf_range = unsafe {
+                                $buf.index_mut_unchecked(($offset.., ..$N))
```

Comment:
Instead of adding `*_unchecked` methods, we can just call `assume` before it.  This works for normal slices as well.  And it feels "safer" since we can have essentially the same condition in the `assume` call here as the `assert!` outside of the `match`, so it's very obvious that it's sound.

Reply:
It's also easy to make soundness mistakes this way, vs. the `assume` way, which just results in unoptimized code if we make a mistake.  I don't think we do this here, but it would be reasonable to call something like this on `Pixels` (so the elements are `BD::Pixel`s), in which case there will be a `* 2` for the index and length calculation.  If one of these overflows, the `unsafe` becomes unsound.  With an `assume`, it just becomes a bounds check.

Reply:
I strongly disagree with this. `assume` is not stable, first, but more importantly even if it was we have a simple API that does exactly what we want (don't insert a check) rather than relying on the optimizer to do something for us.